File: androidApp/src/main/java/io/redlink/more/app/android/shared_composables/ScheduleList.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license (see https://www.apache.org/licenses/LICENSE-2.0).
 */
package io.redlink.more.app.android.shared_composables

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.navigation.NavController
import io.redlink.more.app.android.activities.NavigationScreen
import io.redlink.more.app.android.activities.dashboard.schedule.ScheduleViewModel
import io.redlink.more.app.android.activities.dashboard.schedule.list.ScheduleListItem
import io.redlink.more.app.android.extensions.formattedString
import io.redlink.more.app.android.extensions.jvmLocalDate
import io.redlink.more.navigation.model.NavigationRouteParameter

@Composable
fun ScheduleList(viewModel: ScheduleViewModel, navController: NavController, showButton: Boolean) {
    val schedulesByDate by viewModel.coreViewModel.schedulesByDate.collectAsStateWithLifecycle()

    val sortedScheduleEntries by remember(schedulesByDate) {
        derivedStateOf {
            schedulesByDate.entries
                .filter { it.value.isNotEmpty() }
                .sortedBy { it.key }
                .map { entry ->
                    entry.key to entry.value.sortedWith(
                        compareBy(
                            { it.start },
                            { it.end },
                            { it.observationTitle },
                            { it.scheduleId }
                        )
                    )
                }
        }
    }

    LazyColumn {
        sortedScheduleEntries.forEach { (date, schedules) ->
            item(key = "header_$date") {
                Heading(
                    text = date.jvmLocalDate().formattedString(),
                    modifier = Modifier.fillMaxWidth()
                )
            }

            items(
                items = schedules,
                key = { schedule -> schedule.scheduleId }
            ) { scheduleModel ->
                MoreDivider(Modifier.fillMaxWidth())
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable(
                            indication = null,
                            interactionSource = remember { MutableInteractionSource() }
                        ) {
                            navController.navigate(
                                NavigationScreen.SCHEDULE_DETAILS.navigationRoute(
                                    NavigationRouteParameter.SCHEDULE_ID.key to scheduleModel.scheduleId,
                                    NavigationRouteParameter.SCHEDULE_LIST_TYPE.key to viewModel.scheduleListType
                                )
                            )
                        }
                ) {
                    ScheduleListItem(
                        navController = navController,
                        scheduleModel = { scheduleModel },
                        viewModel = viewModel,
                        showButton = showButton
                    )
                }
            }

            item(key = "spacer_$date") {
                Spacer(modifier = Modifier.height(30.dp))
            }
        }
    }
}

File: iosApp/iosApp/Views/Schedule/List/ScheduleList.swift
//
//  ScheduleList.swift
//  iosApp
//
//  Created by Julia Mayrhauser on 03.03.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung 
//  Licensed under the Apache 2.0 license with Commons Clause 
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import shared
import SwiftUI

struct ScheduleList: View {
    @ObservedObject var viewModel: ScheduleViewModel
    @EnvironmentObject var navigationModalState: NavigationModalState
    var scheduleModels: [ScheduleModel] = []
    var scheduleListType: ScheduleListType
    
    var body: some View {
        ForEach(scheduleModels, id: \.scheduleId) { schedule in
            VStack {
                ScheduleListItem(viewModel: viewModel, scheduleModel: schedule, showButton: scheduleListType != .completed)
                    .environmentObject(navigationModalState)
                Divider()
            }
        }
        
    }
}

struct ScheduleList_Previews: PreviewProvider {
    static var previews: some View {
        ScheduleList(viewModel: ScheduleViewModel(scheduleListType: .all), scheduleModels: [
            ScheduleModel(scheduleId: "id-1", observationId: "observation-id-1", observationType: "type-1", observationTitle: "title-1", done: false, start: 4000000, end: 4500000, hidden: false, scheduleState: .active),
            ScheduleModel(scheduleId: "id-2", observationId: "observation-id-2", observationType: "type-2", observationTitle: "title-2", done: false, start: 4000000, end: 4500000, hidden: false, scheduleState: .active),
        ], scheduleListType: .all)
    }
}
